packages/sources/coinmetrics-test/src/endpoint/price-ws.ts

```diff
@@ -0,0 +1,170 @@
+    frequency: '1s',
+    api_key: context.adapterConfig.API_KEY,
+  }
+  logger.debug(`Generating url with: ${JSON.stringify(params)}`)
+  const generated = new URL(
+    `/v4/timeseries-stream/asset-metrics?${querystring.stringify(params)}`,
+    context.adapterConfig.WS_API_ENDPOINT || DEFAULT_WS_API_ENDPOINT,
+  ).toString()
+  logger.debug(`Generated the following url: ${generated}`)
+  return generated
+}
+
+export const handleAssetMetricsMessage = (
+  message: WsAssetMetricsMessage,
+): ProviderResult<WsAssetMetricsEndpointTypes>[] | undefined => {
+  logger.trace(message, 'Got response from websocket')
+  //Cast message to all of its known types so we can check which variant we received
+  const reorgMessage = message as WsAssetMetricsReorgResponse
+  const warningMessage = message as WsAssetMetricsWarningResponse
+  const successMessage = message as WsAssetMetricsSuccessResponse
+  const errorMessage = message as WsAssetMetricsErrorResponse
+  if (successMessage.asset && successMessage.type !== 'reorg') {
+    // Success message
+    let metric: number | undefined
+    let metricQuote = ''
+    for (const quote of VALID_REFERENCE_RATE_QUOTES) {
+      metric = successMessage[`ReferenceRate${quote}`]
+      if (metric) {
+        metricQuote = quote
+        break
+      }
+    }
+    if (!metric) {
+      //TODO should we throw an error here?
+      logger.warn(
+        successMessage,
+        `Could not find any defined ReferenceRate(${VALID_REFERENCE_RATE_QUOTES.join(
+          '|',
+        )}) properties in message (${JSON.stringify(message)}), will return 0 value as default`,
+      )
+    }
+    return [
+      {
+        params: {
+          base: successMessage.asset,
+          quote: metricQuote,
+          endpoint: 'price-ws',
+        },
+        value: metric || 0, // Fallback zero here just to get past the type checker. We've already checked that metric is defined above.
+      },
+    ]
+  } else if (errorMessage.error) {
+    logger.error(errorMessage, `Error response from websocket`)
+  } else if (warningMessage.warning) {
+    logger.warn(warningMessage, `Warning response from websocket`)
+  } else if (reorgMessage.type === 'reorg') {
+    logger.info(reorgMessage, `Reorg response from websocket`)
+  } else {
+    logger.warn(message, 'Unknown message type from websocket')
+  }
+  return undefined
+}
+
+export const wsTransport = new WebSocketTransport<WsAssetMetricsEndpointTypes>({
+  url: async (context, desiredSubs) => {
+    return await calculateAssetMetricsUrl(context, desiredSubs)
+  },
+  handlers: {
+    open(connection) {
+      return new Promise((resolve) => {
+        logger.debug('Opening websocket connection')
+        logger.debug(connection, 'Connection')
+
+        connection.on('message', (data: WebSocketRawData) => {
+          const parsed = JSON.parse(data.toString())
+          logger.debug(parsed, 'received message on connection')
+          resolve()
+        })
+      })
```

Comment:
If there's no special logic to the open handler for this provider, we can omit it